### ISSUE TITLE
`Stream.groupBy` cannot collect into `HashMap` 

### ISSUE THREAD

**Issue by scarf005** [rimbu-org/rimbu#212]
## Environment

- OS: mac
- runtime: Deno 1.40.5, typescript 5.3.3

## Reproduction

```ts
// test.ts

import { Stream } from "https://deno.land/x/rimbu@1.2.0/stream/mod.ts"
import { HashMap } from "https://deno.land/x/rimbu@1.2.0/hashed/mod.ts"

export const test = (stream: Stream<string>) =>
	stream.groupBy((node) => node, { collector: HashMap.reducer() })

```

## Steps to reproduce

```sh
$ deno check test.ts
```

## Expected Behavior

it type checks.

## Actual Behavior

```ts
error: TS2769 [ERROR]: No overload matches this call.
  Overload 1 of 2, '(valueToKey: (value: string, index: number) => unknown, options: { collector: Reducer<[unknown, string], HashMap<unknown, string>>; }): HashMap<unknown, string>', gave the following error.
    Type 'Reducer<readonly [unknown, string], HashMap<unknown, string>>' is not assignable to type 'Reducer<[unknown, string], HashMap<unknown, string>>'.
      The type 'readonly [unknown, string]' is 'readonly' and cannot be assigned to the mutable type '[unknown, string]'.  Overload 2 of 2, '(valueToKey: (value: string, index: number) => unknown, options?: { collector?: undefined; } | undefined): Map<unknown, string[]>', gave the following error.
    Type 'Reducer<readonly [unknown, string], HashMap<unknown, string>>' is not assignable to type 'undefined'.
	stream.groupBy((node) => node, { collector: HashMap.reducer() })
	                                 ~~~~~~~~~
    at file://<path>/test.ts:5:35

    The expected type comes from property 'collector' which is declared here on type '{ collector: Reducer<[unknown, string], HashMap<unknown, string>>; }'
          collector: Reducer<[K, T], R>;
          ~~~~~~~~~
        at https://deno.land/x/rimbu@1.2.0/stream/main/interface.ts:999:7
```

## Extra context

maybe it's because the type signature for Reducer lacks `readonly` keyword?

https://github.com/rimbu-org/rimbu/blob/3b21319ce2d2ecf3d96bfbb09b688df09405acd0/deno_dist/stream/main/interface.ts#L996-L1001

**Comment by vitoke** [comment]
Ah yes the `readonly` thing still confuses me often. Indeed I think the fix is to make the collector accept a `Reducer<readonly [K, T], R>`

**Comment by scarf005** [comment]
tried again in rimbu 2.0.1, getting this error:

```ts
Stream.of(["a", "b"] as const).reduce(Reducer.toJSObject())
```

```
No overload matches this call.
  Overload 1 of 2, '(reducer: Reducer<readonly ["a", "b"] | ["a", "b"], Record<"a", "b">>): Record<"a", "b">', gave the following error.
    Argument of type 'Reducer<["a", "b"], Record<"a", "b">>' is not assignable to parameter of type 'Reducer<readonly ["a", "b"] | ["a", "b"], Record<"a", "b">>'.
      Types of property 'mapInput' are incompatible.
        Types of parameters 'mapFun' and 'mapFun' are incompatible.
          Type 'readonly ["a", "b"] | ["a", "b"]' is not assignable to type '["a", "b"]'.
            The type 'readonly ["a", "b"]' is 'readonly' and cannot be assigned to the mutable type '["a", "b"]'.  Overload 2 of 2, '(shape: CombineShape<readonly ["a", "b"]>): any', gave the following error.
    Argument of type 'Reducer<["a", "b"], Record<"a", "b">>' is not assignable to parameter of type 'CombineShape<readonly ["a", "b"]>'.
      Type 'Impl<["a", "b"], Record<"a", "b">, unknown>' is not assignable to type 'Reducer<readonly ["a", "b"], any>'.
        Types of property 'mapInput' are incompatible.
          Types of parameters 'mapFun' and 'mapFun' are incompatible.
            The type 'readonly ["a", "b"]' is 'readonly' and cannot be assigned to the mutable type '["a", "b"]'.
```

**Comment by scarf005** [comment]
works well, looks like it's fixed by #214.
